src/dbnode/persist/fs/commitlog/types.go

```diff
@@ -84,26 +84,56 @@ type CommitLog interface {
 	QueueLength() int64
 }
 
+// LogEntry is a commit log entry being read.
+type LogEntry struct {
+	Series     ts.Series
+	Datapoint  ts.Datapoint
+	Unit       xtime.Unit
+	Annotation ts.Annotation
+	Metadata   LogEntryMetadata
+}
+
+// LogEntryMetadata is a set of metadata about a commit log entry being read.
+type LogEntryMetadata struct {
+	// FileReadID is a unique index for the current commit log
+	// file that is being read (only unique per-process).
+	FileReadID uint64
+	// SeriesUniqueIndex is the series unique index relative to the
+	// current commit log file being read.
+	SeriesUniqueIndex uint64
+}
+
 // Iterator provides an iterator for commit logs
 type Iterator interface {
-	// Next returns whether the iterator has the next value
+	// Next returns whether the iterator has the next value.
 	Next() bool
 
-	// Current returns the current commit log entry
-	Current() (ts.Series, ts.Datapoint, xtime.Unit, ts.Annotation)
+	// Current returns the current commit log entry.
+	Current() LogEntry
 
-	// Err returns an error if an error occurred
+	// Err returns an error if an error occurred.
 	Err() error
 
-	// Close the iterator
+	// Close the iterator.
 	Close()
 }
 
 // IteratorOpts is a struct that contains coptions for the Iterator
 type IteratorOpts struct {
-	CommitLogOptions      Options
-	FileFilterPredicate   FileFilterPredicate
-	SeriesFilterPredicate SeriesFilterPredicate
+	CommitLogOptions    Options
+	FileFilterPredicate FileFilterPredicate
+	// ReturnMetadataAsRef will return all series metadata such as ID,
+	// tags and namespace as a reference instead of returning pooled
+	// or allocated byte/string/ID references.
+	// Useful if caller does not hold onto the result between calls to
+	// the next read log entry and wants to avoid allocations and pool
+	// contention.
+	// Note: Series metadata will only be set on the result of a log
+	// entry read if the series is read for the first time for the
+	// combined tuple of FileReadID and SeriesUniqueIndex returned by
+	// the LogEntryMetadata. EncodedTags will also be returned
```

Comment:
Probably not relevant for this PR, but can we add some distinctions to series metadata types as a guard to prevent accidental attempts to access metadata from a series further down the line?